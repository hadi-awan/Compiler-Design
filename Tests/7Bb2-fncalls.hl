// Test functionction definitions and calls
// no parameters, no locals, no recursion
function f() println "call f() and return 0"; return 0; end;
f();

// add parameters
var i;
i = 0;
println "Global i=",i;
function g(i) print "call g(i=",i,") and return i="; return i; end;
g(10);
println "Global i=",i," should still be 0";

// add recursion
function h( i,  n)
  print "call h(",i,",",n,"), "; 
  if i<n then
    return h(i+1,n-1);
  else
    return i;
  fi;   
end;
i=h(1,10);
println " and return ",i;

// add local variables
function a( i,  n)
  var j,k;
  print "call a(",i,",",n,"), "; 
  if i<n then
    j = i+2;
    k = n-1;
    return a(j,k);
  else
    return i;
  fi;   
end;
i = a(1,20);
println " and return ",i;


