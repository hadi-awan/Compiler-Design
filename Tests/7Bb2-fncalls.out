call f() and return 0
0
Global i=0
call g(i=10) and return i=10
Global i=0 should still be 0
call h(1,10), call h(2,9), call h(3,8), call h(4,7), call h(5,6), call h(6,5),  and return 6
call a(1,20), call a(3,19), call a(5,18), call a(7,17), call a(9,16), call a(11,15), call a(13,14), call a(15,13),  and return 15
